File: utils/old/crc32.cpp
#include "crc32.hpp"

namespace utils {
	struct crc32_lookup {
		static const uint32_t Polynomial = 0x82F63B78;
		// CRC-32C
		uint32_t b[ 256 ];
		crc32_lookup() {
			uint32_t i;
			uint32_t j;
			uint32_t crc;
			for( i = 0; i <= 0xFF; i++ ) {
				crc = i;
				for( j = 0; j < 8; j++ ) {
					crc = ( crc >> 1 ) ^ ( - int32_t( crc & 1 ) & Polynomial );
				}
				b[ i ] = crc;
			}
		}
	};
	
	uint32_t crc32( const void* data, int length, uint32_t start ) {
		static const crc32_lookup lookup;
		uint32_t crc = uint32_t( ~start );
		uint8_t const* current = ( uint8_t const* )data;
		while( length-- ) {
			crc = ( crc >> 8 ) ^ lookup.b[ ( crc & 0xFF ) ^ *current++ ];
		}
		return ~crc;
	}
}

File: output/scripts/crc32.lua
local modname = ...
local ffi = require('ffi')

local polynomial = 0x82F63B78 -- CRC-32C
local lookup = ffi.new('uint32_t[256]')
for i = 0, 255 do
	local crc = i
	for j = 0, 7 do
		crc = bit.bxor(bit.rshift(crc, 1), bit.band(-bit.band(crc, 1), polynomial))
	end
	lookup[i] = crc
end

local function crc32(data, length, start)
	local ptr = ffi.cast('uint8_t const*', data)
	if type(data) == 'string' and not length then
		length = #data
	end
	local crc = bit.bnot(start or 0)
	for i = 0, length-1 do
		crc = bit.bxor(bit.rshift(crc, 8), lookup[bit.bxor(bit.band(crc, 255), ptr[i])]);
	end
	return bit.bnot(crc)
end

return crc32
